### PR TITLE
Multiclass decision tree

### PR DESCRIPTION
**Pull request by hectorgon**
@deerzq @yolken @asaluja @Patrick1860 @spencerde

### CODE REVIEW

File: training/src/main/scala/com/airbnb/aerosolve/training/DecisionTreeTrainer.scala [hunk 1]
```diff
@@ -289,6 +317,64 @@ object DecisionTreeTrainer {
...
+    val scale = 1.0 / (sum * sum)
+    var impurity : Double = 0.0
+    for (kv1 <- dist) {
+      for (kv2 <- dist) {
+        if (kv1._1 != kv2._1) {
+          impurity += scale * kv1._2 * kv2._2
+        }
+      }
+    }
+    impurity
```
Comment (deerzq) [inline training/src/main/scala/com/airbnb/aerosolve/training/DecisionTreeTrainer.scala]: multiply `scale` at the end instead of at line327 to save some computation?

File: training/src/main/scala/com/airbnb/aerosolve/training/DecisionTreeTrainer.scala [hunk 2]
```diff
@@ -289,6 +317,64 @@ object DecisionTreeTrainer {
...
+                              splitCriteria : String,
+                              candidateOpt : Option[ModelRecord]): Option[Double] = {
+    val leftDist = scala.collection.mutable.HashMap[String, Double]()
+    val rightDist = scala.collection.mutable.HashMap[String, Double]()
+
+    for (example <- examples) {
+      val response = BoostedStumpsModel.getStumpResponse(candidateOpt.get, example)
+      for (kv <- example.get(rankKey).asScala) {
+        if (response) {
+          rightDist.put(kv._1, kv._2)
```
Comment (deerzq) [inline training/src/main/scala/com/airbnb/aerosolve/training/DecisionTreeTrainer.scala]: I feel the value should be accumulated, otherwise this will be determined by the last example only. how about something like?
`val v = rightDist.getOrElse(kv._1)`

`rightDist.put(kv._1, v + kv._2)`

File: training/src/main/scala/com/airbnb/aerosolve/training/DecisionTreeTrainer.scala [hunk 3]
```diff
@@ -289,6 +317,64 @@ object DecisionTreeTrainer {
...
+        } else {
+          leftDist.put(kv._1, kv._2)
```
Comment (deerzq) [inline training/src/main/scala/com/airbnb/aerosolve/training/DecisionTreeTrainer.scala]: same here

File: training/src/test/scala/com/airbnb/aerosolve/training/DecisionTreeModelTest.scala [hunk 4]
```diff
@@ -44,6 +44,16 @@ class DecisionTreeModelTest {
   }
 
   @Test
+  def testDecisionTreeTrainerMulticlassHellinger() = {
+    testDecisionTreeMulticlassTrainer("multiclass_hellinger", 0.5)
```
Comment (deerzq) [inline training/src/test/scala/com/airbnb/aerosolve/training/DecisionTreeModelTest.scala]: higher threshold?

File: training/src/main/scala/com/airbnb/aerosolve/training/DecisionTreeTrainer.scala (hunk 2 above)
Comment (hectorgon) [inline training/src/main/scala/com/airbnb/aerosolve/training/DecisionTreeTrainer.scala]: Maps are unique associative containers? I don't think it is possible to have more than one value per key in the label map?

File: training/src/test/scala/com/airbnb/aerosolve/training/DecisionTreeModelTest.scala (hunk 4 above)
Comment (hectorgon) [inline training/src/test/scala/com/airbnb/aerosolve/training/DecisionTreeModelTest.scala]: The randomness makes the accuracy jump up and down and could make the test flaky. Since random is 0.25 I figured twice of random should be a good cutoff

File: training/src/main/scala/com/airbnb/aerosolve/training/DecisionTreeTrainer.scala (hunk 2 above)
Comment (deerzq) [inline training/src/main/scala/com/airbnb/aerosolve/training/DecisionTreeTrainer.scala]: `v + kv._2` is just one value, not two values.

File: training/src/main/scala/com/airbnb/aerosolve/training/DecisionTreeTrainer.scala (hunk 1 above)
Comment (hectorgon) [inline training/src/main/scala/com/airbnb/aerosolve/training/DecisionTreeTrainer.scala]: Done

File: training/src/main/scala/com/airbnb/aerosolve/training/DecisionTreeTrainer.scala (hunk 2 above)
Comment (hectorgon) [inline training/src/main/scala/com/airbnb/aerosolve/training/DecisionTreeTrainer.scala]: Ah OK good catch fixed

**Comment by deerzq** [comment]
LGTM!
